src/coreclr/jit/assertionprop.cpp

```diff
@@ -3957,7 +3957,8 @@ void Compiler::optAssertionProp_RangeProperties(ASSERT_VALARG_TP assertions,
         }
 
         // First, analyze possible X ==/!= CNS assertions.
-        if (curAssertion->IsConstantInt32Assertion() && (curAssertion->op1.vn == treeVN))
+        if (curAssertion->IsConstantInt32Assertion() && (curAssertion->op1.kind == O1K_LCLVAR) &&
+            (curAssertion->op1.vn == treeVN))
```

Comment:
Why must the first operand be a LCLVAR to allow using these assertions?

Reply:
@jakobbotsch because `O1K_ARR_BND` is a weird assertion (when used with `OAK_EQUAL/OAK_NOT_EQUAL`). See [here](https://github.com/dotnet/runtime/blob/53cc1ddeec661d03d65d0e2949f3486e2162d80f/src/coreclr/jit/assertionprop.cpp#L2242-L2273)

e.g. when we have `if (arr.Length == 0)` we create an assertion `op1 OAK_NOT_EQUAL 0` (inside `true` edge). To be fair, I think this one should be removed completely and be replaced with `O1K_LCLVAR` (not sure it works for trees not being `GT_LCLVAR`).